### PR TITLE
Use latest dev dependencies

### PR DESCRIPTION
**Pull request by iainelder**
Pin a minimum version for the dev dependencies to ensure sane behavior of Poetry's resolver.

Commit 8bbe6217 removed the version constraints from all dependencies in `pyproject.toml`. It's the right thing to do for the application dependencies, but it's counter-productive for the dev dependencies.

flakeheaven stopped working yesterday. Poetry selected prehistoric flakeheaven version 0.11.0. The [changelog](https://github.com/flakeheaven/flakeheaven/blob/4591fd3dd04c8180d702e5eab3ec25947d7d3393/CHANGELOG.md) stops at 0.11.1!

```console
$ poetry install
Creating virtualenv botocove-r3vy5Ybn-py3.8 in /home/isme/.cache/pypoetry/virtualenvs
Updating dependencies
Resolving dependencies... (11.0s)

Package operations: 76 installs, 0 updates, 0 removals

  ...
  • Installing flakeheaven (0.11.0)
  ...

Writing lock file

Installing the current project: botocove (1.7.3)
```

```console
$ pre-commit run flakeheaven --all-files
flakeheaven..............................................................Failed
- hook id: flakeheaven
- exit code: 1

Traceback (most recent call last):
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/bin/flakeheaven", line 5, in <module>
    from flakeheaven import entrypoint
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/__init__.py", line 5, in <module>
    from ._cli import entrypoint, flake8_entrypoint
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/_cli.py", line 9, in <module>
    from .commands import COMMANDS
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/commands/__init__.py", line 5, in <module>
    from ._baseline import baseline_command
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/commands/_baseline.py", line 6, in <module>
    from .._patched import FlakeHeavenApplication
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/_patched/__init__.py", line 2, in <module>
    from ._app import FlakeHeavenApplication
  File "/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flakeheaven/_patched/_app.py", line 10, in <module>
    from flake8.options.config import ConfigParser, get_local_plugins
ImportError: cannot import name 'ConfigParser' from 'flake8.options.config' (/home/isme/.cache/pypoetry/virtualenvs/botocove-r3vy5Ybn-py3.8/lib/python3.8/site-packages/flake8/options/config.py)
```

The latest CI run ([#73](https://github.com/iainelder/botocove/actions/runs/6443200026/job/17494862297#logs)) in my project fork started at 2023-10-07T20:06Z. That run installed flakeheaven 3.3.0 and the pre-commit check passed.

> Install dev packages
>
> ```text
>   • Installing flakeheaven (3.3.0)
> ```
>
> Lint with flakeheaven
>
> ```text
> flakeheaven..............................................................Passed
> ```

### CODE REVIEW

File: pyproject.toml
Comment (iainelder) [inline pyproject.toml:34]: Because flakeheaven is stuck on flake8 4, it's now a drag on the rest of the flake8-based linters. We should look for a replacement.

The [flakeheaven ex-maintainer](https://github.com/flakeheaven/flakeheaven/issues/132#issuecomment-1752011839) recommends [ruff](https://github.com/astral-sh/ruff), but it may replace flake8 entirely, so it might require some reconfiguration.

**Comment by iainelder** [comment]
[My dev notes for this change](https://github.com/iainelder/botocove-dev-notes/blob/main/dev-notes/Fix-flakeheaven-error.md).

**Comment by iainelder** [comment]
Until you merge the PR I can use this command on another branch to install a working version of flakeheaven.

```bash
bash -c '
    set -e
    rm -rf poetry.lock "$(poetry env info --path)"
    poetry install
    poetry update flakeheaven
    pre-commit run flakeheaven --all-files
'
```

The last two commands give this output:

```text
Updating dependencies
Resolving dependencies... (2.4s)

Package operations: 0 installs, 7 updates, 0 removals

  • Downgrading mccabe (0.7.0 -> 0.6.1)
  • Downgrading pycodestyle (2.9.1 -> 2.8.0)
  • Downgrading pyflakes (2.5.0 -> 2.4.0)
  • Downgrading flake8 (5.0.4 -> 4.0.1)
  • Downgrading flake8-eradicate (1.5.0 -> 1.4.0)
  • Updating flakeheaven (0.11.0 -> 3.3.0)
  • Downgrading pep8-naming (0.13.3 -> 0.13.2)

Writing lock file
flakeheaven..............................................................Passed
```

**Comment by connelldave** [comment]
Thanks for this - I'm happy to move to Ruff but will merge this now to unbreak. I can only roughly remember why flakeheaven is there in the first place, something about flakehell didn't work with something but fixed something else from flake8

**Review by connelldave — APPROVED** [review]

**Comment by iainelder** [comment]
Before trying ruff I'd check whether we can now use flake8 6 without flakeheaven.

**Pull request merged**
